### PR TITLE
KSQL-235:  dont block on socket read

### PR DESCRIPTION
**Pull request by rodesai**
KSQL-235: dont block on socket read

Avoid blocking on socket reads from the cli when
streaming a query response. The read cannot be interrupted,
so the user cannot abort the query via ctl-c until then next
response chunk arrives. Instead, have read poll the input
stream periodically and sleep interruptably between poll
attempts.

### CODE REVIEW

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java [hunk 1]
```diff
@@ -201,11 +203,37 @@ private Response makeGetRequest(String path) {
     public QueryStream(Response response) {
       this.response = response;
 
+      this.closed = false;
       this.objectMapper = new ObjectMapper();
-      this.responseScanner = new Scanner((InputStream) response.getEntity(), StandardCharsets.UTF_8.name());
+      InputStreamReader isr = new InputStreamReader((InputStream)response.getEntity(), StandardCharsets.UTF_8);
+      Object lk = this;
```
Comment (dguy) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java]: nit: `lock`

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java [hunk 2]
```diff
@@ -256,15 +284,22 @@ public void close() {
...
+        closed = true;
+        this.notifyAll();
+      }
       responseScanner.close();
       response.close();
     }
 
-    private IllegalStateException closedIllegalStateException(String methodName) {
-      return new IllegalStateException("Cannot call " + methodName + " when QueryStream is closed");
+    private String closedIllegalStateMsg(String methodName) {
```
Comment (dguy) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java]: Don't see why this method was extracted? Could just call the method below with the methodName

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java [hunk 3]
```diff
@@ -201,11 +203,37 @@ private Response makeGetRequest(String path) {
     public QueryStream(Response response) {
       this.response = response;
 
+      this.closed = false;
       this.objectMapper = new ObjectMapper();
-      this.responseScanner = new Scanner((InputStream) response.getEntity(), StandardCharsets.UTF_8.name());
+      InputStreamReader isr = new InputStreamReader((InputStream)response.getEntity(), StandardCharsets.UTF_8);
+      Object lk = this;
+      String errMsg = closedIllegalStateMsg("hasNext()");
+      this.responseScanner = new Scanner(new Readable() {
```
Comment (dguy) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java]: lambda?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java [hunk 4]
```diff
@@ -201,11 +203,37 @@ private Response makeGetRequest(String path) {
...
+        public int read(CharBuffer buf) throws IOException {
+          int wait = 1;
+          // poll the input stream's readiness between interruptable sleeps
+          // this ensures we cannot block indefinitely on read()
+          while (true) {
+            if (closed) throw new IllegalStateException(errMsg);
```
Comment (dguy) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java]: why are we doing this check outside the `synchronized`? if `closed` is not volatile there is a chance this thread won't see the change. I think it would be better to move into the `synchronized` and removed the double check

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java [hunk 5]
```diff
@@ -201,11 +203,37 @@ private Response makeGetRequest(String path) {
     public QueryStream(Response response) {
       this.response = response;
 
+      this.closed = false;
```
Comment (dguy) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java]: `closed` should be volatile if it is read/updated from multiple threads unless it is always surrounded with `synchronized(...)`

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java [hunk 6]
```diff
@@ -76,11 +77,71 @@ public void testKsqlResource() {
 
 
   @Test
-  public void testStreamQuery() {
+  public void testStreamQuery() throws InterruptedException {
```
Comment (dguy) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: What is this testing? The name doesn't tell me much and The test is hard to follow.

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java [hunk 7]
```diff
@@ -76,11 +77,71 @@ public void testKsqlResource() {
...
+    Assert.assertEquals(1, row.getColumns().
+            size());
+    Assert.assertEquals("hello", row.getColumns().
+            get(0));
+    writer.finished();
+  }
+
+  @Test
+  public void testInterruptStreamedQuery() throws InterruptedException {
+    MockStreamedQueryResource sqr = mockApplication.getStreamedQueryResource();
```
Comment (dguy) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: The first part of this test looks identical to the first part of the test above, i.e, same assertions. No need to assert the same stuff again as it has been proven to work already.

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java [hunk 8]
```diff
@@ -76,11 +77,71 @@ public void testKsqlResource() {
...
+    GenericRow row = sr.getRow();
+    Assert.assertEquals(1, row.getColumns().
+            size());
+    Assert.assertEquals("hello", row.getColumns().
+            get(0));
+    writer.finished();
+  }
+
+  @Test
+  public void testInterruptStreamedQuery() throws InterruptedException {
```
Comment (dguy) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: `shouldInterruptScannerOnClose`

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java [hunk 9]
```diff
@@ -76,11 +77,71 @@ public void testKsqlResource() {
...
+    MockStreamedQueryResource.TestStreamWriter writer = writers.get(0);
+    writer.enq("hello");
+    KsqlRestClient.QueryStream queryStream = queryResponse.getResponse();
+    Assert.assertTrue(queryStream.hasNext());
+    StreamedRow sr = queryStream.next();
+    Assert.assertNotNull(sr);
+    GenericRow row = sr.getRow();
+    Assert.assertEquals(1, row.getColumns().size());
+    Assert.assertEquals("hello", row.getColumns().get(0));
+    List<Boolean> threw = new java.util.LinkedList<>();
```
Comment (dguy) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: `CountDownLatch`?

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java [hunk 10]
```diff
@@ -76,11 +77,71 @@ public void testKsqlResource() {
...
+      public void run() {
+        try {
+          queryStream.hasNext();
+        } catch (IllegalStateException e) {
+          threw.set(0, true);
+        }
+      }
+    };
+    t.start();
+    Thread.sleep(100);
```
Comment (dguy) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: remove this and wait on a `CountDownLatch`

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java [hunk 11]
```diff
@@ -76,11 +77,71 @@ public void testKsqlResource() {
...
+    queryStream.close();
+    t.join(10000);
+    Assert.assertTrue(threw.get(0));
+    writer.finished();
```
Comment (dguy) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: if this is required then it should either be done in `tearDown` or in a `finally` block - same above

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java (hunk 3 above)
Comment (rodesai) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java]: For some reason I thought we had to support building on Java 7. I'll change it.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java (hunk 4 above)
Comment (rodesai) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java]: Thanks. The check is there so that we exit even if a line is ready after being interrupted. I'll move it into the synchronized block.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java (hunk 2 above)
Comment (rodesai) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java]: java misunderstanding on my end. fixing.

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java (hunk 6 above)
Comment (rodesai) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: It verifies that we can receive a row sent by the server. I'll update the name and add some comments

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java [hunk 12]
```diff
@@ -76,11 +79,74 @@ public void testKsqlResource() {
...
+          queryStream.hasNext();
+        } catch (IllegalStateException e) {
+          threw.countDown();
+        }
+      });
+      t.setDaemon(true);
+      t.start();
+
+      // Let the thread run and then close the stream. Verify that it was interrupted
+      Thread.sleep(100);
```
Comment (dguy) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: rather than sleeping wait on the latch, i.e,
`assertTrue(threw.await(100))`

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java [hunk 13]
```diff
@@ -76,11 +79,74 @@ public void testKsqlResource() {
...
+      queryStream.close();
+      Assert.assertTrue(threw.await(10, TimeUnit.SECONDS));
```
Comment (dguy) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: this line but instead of the `thread.sleep(100)`

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java (hunk 12 above)
Comment (rodesai) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: The intent behind the sleep was to give the thread a chance to run and block so that it gets interrupted when close is called. That probably won't happen if I call close immediately and then wait on the latch.

Comment (dguy) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: actually what i think you should do here is immediately close the `QueryStream` after line 129. Then the first call to `queryStream.hasNext()` will throw the expected exception. You can remove the sleep and just wait on the latch

Comment (rodesai) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/client/KsqlRestClientTest.java]: I understand. My objection is that I want the query thread to actually sleep, so that I can verify that it behaves correctly after being interrupted. Implementing the test the way you are suggesting will result in the query thread throwing the exception before it sleeps. That said, I'm not a huge fan of timing-based tests. But I don't see a better way of doing it other than instrumenting QueryStream just for the sake of testing this case.

**Comment by rodesai** [comment]
retest this please

**Comment by rodesai** [comment]
retest this please
